File: src/dataset/javascript/luhn_exercism-javascript-ffflorian/openAI/ChatGPT4_luhn_v4.js
//@ts-check

/**
 * @param {string} ccnumber
 * @returns {boolean}
 */
export const valid = ccnumber => {
  if (!/^\d[\d ]+$/.test(ccnumber)) return false;

  let sum = 0;
  let shouldDouble = false;

  for (let i = ccnumber.length - 1; i >= 0; i--) {
    const char = ccnumber[i];
    if (char === ' ') continue;

    let digit = char.charCodeAt(0) - 48;
    if (shouldDouble) {
      digit = digit * 2;
      if (digit > 9) digit -= 9;
    }

    sum += digit;
    shouldDouble = !shouldDouble;
  }

  return sum % 10 === 0;
};

File: src/dataset/go/luhn_Exercism_go_rootulp/openAI/ChatGPT4_luhn_v4.go
package luhn

import (
	"errors"
	"strconv"
)

// Valid returns whether the provided string represents a number that satisfies the Luhn algorithm.
func Valid(s string) bool {
	n := len(s)
	if n <= 1 {
		return false
	}

	sum, double := 0, false
	for i := n - 1; i >= 0; i-- {
		if s[i] == ' ' {
			continue
		}

		digit := int(s[i] - '0')
		if digit < 0 || digit > 9 {
			return false
		}

		if double {
			digit *= 2
			if digit > 9 {
				digit -= 9
			}
		}
		sum += digit
		double = !double
	}

	return sum%10 == 0
}

// checkSum returns the Luhn check sum of the provided string.
func checkSum(s string) (int, error) {
	n := len(s)
	sum, double := 0, false

	for i := n - 1; i >= 0; i-- {
		if s[i] == ' ' {
			continue
		}

		digit := int(s[i] - '0')
		if digit < 0 || digit > 9 {
			return 0, errors.New("Invalid digit")
		}

		if double {
			digit *= 2
			if digit > 9 {
				digit -= 9
			}
		}
		sum += digit
		double = !double
	}

	return sum, nil
}